### PR TITLE
Refactor architecture of prompts for better key handling

### PR DESCRIPTION
**Pull request by mikaelmello**
Major architecture changes:

1. Introduction of Action enums as an intermediary layer between Key events and the prompt's business logic.

There's a top level Action where one of its variants is Inner, with a member of a generic type.

2. Introduction of Config structure for each prompts.

This is still very underutilized. I needed to export some fields to an external struct for proper key to action handling in some cases that depended on the prompt config (e.g. vim key bindings)

In the future, it might be worthwhile to explore using solely this struct for prompt specific fields and then moving shared behavior to a common Prompt struct.

3. Prompt trait

All prompts implement this top level trait which implements a couple of methods that were basically duplicated code before.

4. Input actions

The input struct now also handles actions instead of key events.

---

Future steps after this design change will include exposing calling prompts with a predefined list of actions, instead of relying on reading keys from the terminal.

### CODE REVIEW

**Comment by mikaelmello** [comment]
@IniterWorker, I was thinking of merging this later today.

Even though there are tons of changes, the underlying API has been basically untouched. New modules are public but there are no major nor minor breaking changes.

If you'd be interested in taking a look, I'd be happy to hear your thoughts on the changes I've made

File: inquire/src/date_utils.rs [hunk 1]
```diff
@@ -3,11 +3,11 @@ use core::panic;
 use chrono::NaiveDate;
 
 pub fn get_current_date() -> NaiveDate {
-    chrono::Local::now().date().naive_local()
+    chrono::Local::now().date_naive()
 }
 
 pub fn get_start_date(month: chrono::Month, year: i32) -> NaiveDate {
-    chrono::NaiveDate::from_ymd(year, month.number_from_month(), 1)
+    chrono::NaiveDate::from_ymd_opt(year, month.number_from_month(), 1).unwrap()
```
Comment (IniterWorker) [inline inquire/src/date_utils.rs:10]: Maybe, we should add `expect("start_data: out-of-range date, invalid month and/or day")` to at least document the corner case.

```suggestion
    chrono::NaiveDate::from_ymd_opt(year, month.number_from_month(), 1).expect("start_data: out-of-range date, invalid month and/or day")
```

**Review by IniterWorker — APPROVED** [review]
__Mikael Mello__, your work is very promissing. It will drastically improve our capabilities to test the `inquire` library.

Of course, this approach brings complexity to the derive's design. I need to revise my approach to reflect your changes to allow a custom `backend`.

I think we could improve the helper functions like generating KeyEvent or KeyCode by making them public. We note the desire of some users to want to test their entry. What do you think ?
- Less code duplication/documentation.
- Allows the user to test their inputs.

```rust
macro_rules! text_to_events {
    ($text:expr) => {{
        $text.chars().map(KeyCode::Char)
    }};
}
```

Great work!

**Comment by mikaelmello** [comment]
> I think we could improve the helper functions like generating KeyEvent or KeyCode by making them public. We note the desire of some users to want to test their entry. What do you think ?

Fully agree, in my mind I had as the next step to expose a `prompt()` method that receives a list of actions, for the testability aspect you mentioned.

> Of course, this approach brings complexity to the derive's design. I need to revise my approach to reflect your changes to allow a custom backend.

The new Prompt trait is private and I was thinking of keeping it like that, being it sort of an implementation detail. This means that all prompts' interfaces haven't changed, custom backends included (still crate-only). I think we can get away without needing any external changes

**Pull request merged**